cdc/cdc.cc

```diff
@@ -0,0 +1,197 @@
+    auto& replication_strategy =
+        db.find_keyspace(s.ks_name()).get_replication_strategy();
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    auto ignore_msb_bits =
+        db.get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto belongs_to = [&](const gms::inet_address& endpoint,
+                          const unsigned int shard_id,
+                          const int shard_count,
+                          const utils::UUID& stream_id) {
+        const auto log_pk = partition_key::from_singular(*log_schema,
+                                                         data_value(stream_id));
+        const auto token =
+            dht::global_partitioner().decorate_key(*log_schema, log_pk).token();
+        auto&& eps =
+            replication_strategy.calculate_natural_endpoints(token, token_metadata);
+        if (eps.empty() || eps[0] != endpoint) {
+            return false;
+        }
+        const auto owning_shard_id = dht::murmur3_partitioner::shard_of(
+                token, shard_count, ignore_msb_bits);
+        return owning_shard_id == shard_id;
+    };
+
+    std::vector<mutation> mutations;
+    const auto ts = api::new_timestamp();
+    const auto ck = clustering_key::from_single_value(
+            *desc_schema, timestamp_type->decompose(ts));
+    auto cdef = desc_schema->get_column_definition(to_bytes("stream_id"));
+
+    for_each_endpoint([&] (const gms::inet_address& endpoint) {
+        const auto decomposed_ip = inet_addr_type->decompose(endpoint.addr());
+
+        for_each_shard(endpoint, [&] (unsigned int shard_id, int shard_count) {
+            const auto pk = partition_key::from_exploded(
+                    *desc_schema, { decomposed_ip, int32_type->decompose(static_cast<int>(shard_id)) });
+            mutations.emplace_back(desc_schema, pk);
+
+            auto stream_id = utils::make_random_uuid();
+            while (!belongs_to(endpoint, shard_id, shard_count, stream_id)) {
+                stream_id = utils::make_random_uuid();
+            }
+            auto value = atomic_cell::make_live(*uuid_type,
+                                                ts,
+                                                uuid_type->decompose(stream_id));
+            mutations.back().set_cell(ck, *cdef, std::move(value));
+        });
+    });
+    return proxy.mutate(std::move(mutations),
+                        db::consistency_level::QUORUM,
+                        db::no_timeout,
+                        nullptr,
+                        empty_service_permit());
+}
+
+future<> setup(service::storage_proxy& proxy, schema_ptr s) {
+    return setup_log(*s).then([&proxy, s = std::move(s)] () mutable {
+        return setup_desc(*s).then_wrapped([&proxy, s = std::move(s)] (future<> f) mutable {
```

Comment:
`announce_new_column_family` might throw.

If that happens inside `setup_desc`, it will fly out of `setup_desc` and make `setup` return an exceptional future. But whatever happened in `setup_log` won't be reversed.
You might want to wrap `setup_desc` in a `make_ready_future<>().then([] {... setup_desc ...})` construct.

If that happens inside `setup_log`, `setup` will propagate the exception too. Perhaps we want to return an exceptional future instead.